apps/meteor/app/e2e/server/functions/fetchUsersWaitingForGroupKey.ts

```diff
@@ -0,0 +1,46 @@
+import type { IRoom, IUser } from '@rocket.chat/core-typings';
+import { Rooms, Subscriptions, Users } from '@rocket.chat/models';
+
+import { isTruthy } from '../../../../lib/isTruthy';
+
+const NUMBER_OF_ROOMS = 10;
+const QUEUE_SIZE = 50;
+
+export async function fetchUsersWaitingForGroupKey(userId: IUser['_id']): Promise<{
+	usersWaitingForE2EKeys: Record<IRoom['_id'], { _id: IUser['_id']; public_key: string }[]>;
+	hasMore: boolean;
+}> {
+	const roomIds = (await Subscriptions.findByUserId(userId, { projection: { rid: 1 } }).toArray()).map((sub) => sub.rid);
+
+	const cursor = Rooms.findRoomsByRoomIdsWithE2EEQueue(roomIds, { projection: { usersWaitingForE2EKeys: 1 } });
```

Comment:
Got it

Reply:
Why not use the sample here to return the random ids already? It can be limited to `NUMBER_OF_ROOMS +1`, so if the result is `> NUMBER_OF_ROOMS` just remove the last item from the array and define hasMore as true.